File: player-core/src/main/java/com/libreshockwave/player/debug/BreakpointManager.java
package com.libreshockwave.player.debug;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Manages breakpoints with support for enable/disable.
 * Thread-safe for concurrent access from UI and VM threads.
 */
public class BreakpointManager {

    /**
     * Key for breakpoint lookup (scriptId:offset).
     */
    public record BreakpointKey(int scriptId, int offset) {
        public static BreakpointKey of(int scriptId, int offset) {
            return new BreakpointKey(scriptId, offset);
        }

        public static BreakpointKey of(Breakpoint bp) {
            return new BreakpointKey(bp.scriptId(), bp.offset());
        }

        @Override
        public String toString() {
            return scriptId + ":" + offset;
        }
    }

    private final ConcurrentHashMap<BreakpointKey, Breakpoint> breakpoints = new ConcurrentHashMap<>();

    /**
     * Get a breakpoint at the specified location, or null if none exists.
     */
    public Breakpoint getBreakpoint(int scriptId, int offset) {
        return breakpoints.get(BreakpointKey.of(scriptId, offset));
    }

    /**
     * Check if a breakpoint exists at the specified location.
     */
    public boolean hasBreakpoint(int scriptId, int offset) {
        return breakpoints.containsKey(BreakpointKey.of(scriptId, offset));
    }

    /**
     * Set or update a breakpoint.
     */
    public void setBreakpoint(Breakpoint bp) {
        breakpoints.put(BreakpointKey.of(bp), bp);
    }

    /**
     * Add a simple breakpoint. If one already exists, it will be replaced.
     */
    public Breakpoint addBreakpoint(int scriptId, int offset) {
        Breakpoint bp = Breakpoint.simple(scriptId, offset);
        setBreakpoint(bp);
        return bp;
    }

    /**
     * Remove a breakpoint.
     * @return the removed breakpoint, or null if none existed
     */
    public Breakpoint removeBreakpoint(int scriptId, int offset) {
        return breakpoints.remove(BreakpointKey.of(scriptId, offset));
    }

    /**
     * Toggle a breakpoint at the specified location.
     * If no breakpoint exists, creates a simple enabled breakpoint.
     * If a breakpoint exists, removes it.
     * @return the new breakpoint if added, or null if removed
     */
    public Breakpoint toggleBreakpoint(int scriptId, int offset) {
        BreakpointKey key = BreakpointKey.of(scriptId, offset);
        Breakpoint existing = breakpoints.get(key);
        if (existing != null) {
            breakpoints.remove(key);
            return null;
        } else {
            Breakpoint bp = Breakpoint.simple(scriptId, offset);
            breakpoints.put(key, bp);
            return bp;
        }
    }

    /**
     * Toggle the enabled state of a breakpoint.
     * @return the updated breakpoint, or null if no breakpoint exists
     */
    public Breakpoint toggleEnabled(int scriptId, int offset) {
        BreakpointKey key = BreakpointKey.of(scriptId, offset);
        return breakpoints.computeIfPresent(key, (k, bp) -> bp.withEnabled(!bp.enabled()));
    }

    /**
     * Get all breakpoints.
     */
    public Collection<Breakpoint> getAllBreakpoints() {
        return Collections.unmodifiableCollection(breakpoints.values());
    }

    /**
     * Get all breakpoints for a specific script.
     */
    public List<Breakpoint> getBreakpointsForScript(int scriptId) {
        return breakpoints.values().stream()
            .filter(bp -> bp.scriptId() == scriptId)
            .collect(Collectors.toList());
    }

    /**
     * Get all offsets with breakpoints for a specific script.
     */
    public Set<Integer> getOffsetsForScript(int scriptId) {
        return breakpoints.values().stream()
            .filter(bp -> bp.scriptId() == scriptId)
            .map(Breakpoint::offset)
            .collect(Collectors.toSet());
    }

    /**
     * Clear all breakpoints.
     */
    public void clearAll() {
        breakpoints.clear();
    }

    /**
     * Get all breakpoints as a map (scriptId -> set of offsets).
     * This is for backward compatibility with the old breakpoint storage format.
     */
    public Map<Integer, Set<Integer>> toOffsetMap() {
        Map<Integer, Set<Integer>> result = new HashMap<>();
        for (Breakpoint bp : breakpoints.values()) {
            result.computeIfAbsent(bp.scriptId(), k -> new HashSet<>()).add(bp.offset());
        }
        return result;
    }

    /**
     * Set breakpoints from an offset map (old format).
     * Creates simple enabled breakpoints for each entry.
     */
    public void setFromOffsetMap(Map<Integer, Set<Integer>> offsetMap) {
        clearAll();
        if (offsetMap != null) {
            for (Map.Entry<Integer, Set<Integer>> entry : offsetMap.entrySet()) {
                int scriptId = entry.getKey();
                for (int offset : entry.getValue()) {
                    setBreakpoint(Breakpoint.simple(scriptId, offset));
                }
            }
        }
    }

    /**
     * Serialize breakpoints to JSON format.
     */
    public String serialize() {
        if (breakpoints.isEmpty()) {
            return "";
        }

        StringBuilder sb = new StringBuilder();
        sb.append("{\"version\":2,\"breakpoints\":[");

        boolean first = true;
        for (Breakpoint bp : breakpoints.values()) {
            if (!first) sb.append(",");
            first = false;
            sb.append("{");
            sb.append("\"scriptId\":").append(bp.scriptId()).append(",");
            sb.append("\"offset\":").append(bp.offset()).append(",");
            sb.append("\"enabled\":").append(bp.enabled());
            sb.append("}");
        }

        sb.append("]}");
        return sb.toString();
    }

    /**
     * Deserialize breakpoints from a string.
     * Supports both new JSON format and legacy format for backward compatibility.
     */
    public void deserialize(String data) {
        clearAll();
        if (data == null || data.isEmpty()) {
            return;
        }

        data = data.trim();
        if (data.startsWith("{")) {
            // New JSON format
            deserializeJson(data);
        } else {
            // Legacy format: "scriptId:offset,offset;scriptId:offset,offset;..."
            deserializeLegacy(data);
        }
    }

    private void deserializeJson(String json) {
        try {
            // Simple JSON parsing (no external dependencies)
            // Expected format: {"version":2,"breakpoints":[{...},{...}]}

            // Find breakpoints array
            int breakpointsStart = json.indexOf("\"breakpoints\":[");
            if (breakpointsStart < 0) return;

            int arrayStart = json.indexOf('[', breakpointsStart);
            int arrayEnd = json.lastIndexOf(']');
            if (arrayStart < 0 || arrayEnd < 0 || arrayEnd <= arrayStart) return;

            String arrayContent = json.substring(arrayStart + 1, arrayEnd);
            if (arrayContent.isBlank()) return;

            // Parse each breakpoint object
            int depth = 0;
            int objStart = -1;
            for (int i = 0; i < arrayContent.length(); i++) {
                char c = arrayContent.charAt(i);
                if (c == '{') {
                    if (depth == 0) objStart = i;
                    depth++;
                } else if (c == '}') {
                    depth--;
                    if (depth == 0 && objStart >= 0) {
                        String objStr = arrayContent.substring(objStart, i + 1);
                        Breakpoint bp = parseBreakpointJson(objStr);
                        if (bp != null) {
                            setBreakpoint(bp);
                        }
                        objStart = -1;
                    }
                }
            }
        } catch (Exception e) {
            // On parse error, try legacy format as fallback
            deserializeLegacy(json);
        }
    }

    private Breakpoint parseBreakpointJson(String json) {
        try {
            int scriptId = parseJsonInt(json, "scriptId", 0);
            int offset = parseJsonInt(json, "offset", 0);
            boolean enabled = parseJsonBoolean(json, "enabled", true);

            return new Breakpoint(scriptId, offset, enabled);
        } catch (Exception e) {
            return null;
        }
    }

    private int parseJsonInt(String json, String key, int defaultValue) {
        String pattern = "\"" + key + "\":";
        int idx = json.indexOf(pattern);
        if (idx < 0) return defaultValue;

        int start = idx + pattern.length();
        int end = start;
        while (end < json.length() && (Character.isDigit(json.charAt(end)) || json.charAt(end) == '-')) {
            end++;
        }
        if (end == start) return defaultValue;

        try {
            return Integer.parseInt(json.substring(start, end));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    private boolean parseJsonBoolean(String json, String key, boolean defaultValue) {
        String pattern = "\"" + key + "\":";
        int idx = json.indexOf(pattern);
        if (idx < 0) return defaultValue;

        int start = idx + pattern.length();
        if (json.regionMatches(start, "true", 0, 4)) return true;
        if (json.regionMatches(start, "false", 0, 5)) return false;
        return defaultValue;
    }

    private void deserializeLegacy(String data) {
        // Format: "scriptId:offset,offset;scriptId:offset,offset;..."
        try {
            String[] scripts = data.split(";");
            for (String script : scripts) {
                if (script.isEmpty()) continue;
                String[] parts = script.split(":");
                if (parts.length != 2) continue;
                int scriptId = Integer.parseInt(parts[0]);
                String[] offsetStrs = parts[1].split(",");
                for (String offsetStr : offsetStrs) {
                    if (!offsetStr.isEmpty()) {
                        int offset = Integer.parseInt(offsetStr);
                        setBreakpoint(Breakpoint.simple(scriptId, offset));
                    }
                }
            }
        } catch (NumberFormatException e) {
            // Ignore parse errors in legacy format
        }
    }

    /**
     * Serialize breakpoints to legacy format for backward compatibility.
     * Format: "scriptId:offset,offset;scriptId:offset,offset;..."
     */
    public String serializeLegacy() {
        Map<Integer, Set<Integer>> offsetMap = toOffsetMap();
        StringBuilder sb = new StringBuilder();
        boolean first = true;
        for (Map.Entry<Integer, Set<Integer>> entry : offsetMap.entrySet()) {
            if (entry.getValue().isEmpty()) continue;
            if (!first) sb.append(";");
            first = false;
            sb.append(entry.getKey()).append(":");
            boolean firstOffset = true;
            for (Integer offset : entry.getValue()) {
                if (!firstOffset) sb.append(",");
                firstOffset = false;
                sb.append(offset);
            }
        }
        return sb.toString();
    }
}

File: cpp/src/player/debug/BreakpointManager.cpp
#include "libreshockwave/player/debug/BreakpointManager.hpp"

#include <cctype>
#include <sstream>
#include <utility>

namespace libreshockwave::player::debug {
namespace {

std::string trim(std::string value) {
    auto begin = value.begin();
    while (begin != value.end() && std::isspace(static_cast<unsigned char>(*begin))) {
        ++begin;
    }
    auto end = value.end();
    while (end != begin && std::isspace(static_cast<unsigned char>(*(end - 1)))) {
        --end;
    }
    return std::string(begin, end);
}

std::vector<std::string> split(const std::string& value, char delimiter) {
    std::vector<std::string> parts;
    std::string current;
    std::istringstream in(value);
    while (std::getline(in, current, delimiter)) {
        parts.push_back(current);
    }
    return parts;
}

} // namespace

std::optional<Breakpoint> BreakpointManager::getBreakpoint(int scriptId,
                                                           const std::string& handlerName,
                                                           int offset) const {
    const auto it = breakpoints_.find(BreakpointKey::of(scriptId, handlerName, offset));
    return it == breakpoints_.end() ? std::nullopt : std::optional<Breakpoint>(it->second);
}

bool BreakpointManager::hasBreakpoint(int scriptId, const std::string& handlerName, int offset) const {
    return breakpoints_.contains(BreakpointKey::of(scriptId, handlerName, offset));
}

void BreakpointManager::setBreakpoint(Breakpoint breakpoint) {
    breakpoints_[BreakpointKey::of(breakpoint)] = std::move(breakpoint);
}

Breakpoint BreakpointManager::addBreakpoint(int scriptId, std::string handlerName, int offset) {
    Breakpoint breakpoint = Breakpoint::simple(scriptId, std::move(handlerName), offset);
    setBreakpoint(breakpoint);
    return breakpoint;
}

std::optional<Breakpoint> BreakpointManager::removeBreakpoint(int scriptId,
                                                              const std::string& handlerName,
                                                              int offset) {
    const auto key = BreakpointKey::of(scriptId, handlerName, offset);
    const auto it = breakpoints_.find(key);
    if (it == breakpoints_.end()) {
        return std::nullopt;
    }
    Breakpoint removed = it->second;
    breakpoints_.erase(it);
    return removed;
}

std::optional<Breakpoint> BreakpointManager::toggleBreakpoint(int scriptId,
                                                              std::string handlerName,
                                                              int offset) {
    const auto key = BreakpointKey::of(scriptId, handlerName, offset);
    const auto it = breakpoints_.find(key);
    if (it != breakpoints_.end()) {
        breakpoints_.erase(it);
        return std::nullopt;
    }
    Breakpoint breakpoint = Breakpoint::simple(scriptId, std::move(handlerName), offset);
    breakpoints_[BreakpointKey::of(breakpoint)] = breakpoint;
    return breakpoint;
}

std::optional<Breakpoint> BreakpointManager::toggleEnabled(int scriptId,
                                                           const std::string& handlerName,
                                                           int offset) {
    const auto key = BreakpointKey::of(scriptId, handlerName, offset);
    const auto it = breakpoints_.find(key);
    if (it == breakpoints_.end()) {
        return std::nullopt;
    }
    it->second = it->second.withEnabled(!it->second.enabled);
    return it->second;
}

std::vector<Breakpoint> BreakpointManager::getAllBreakpoints() const {
    std::vector<Breakpoint> result;
    for (const auto& [key, breakpoint] : breakpoints_) {
        (void)key;
        result.push_back(breakpoint);
    }
    return result;
}

std::vector<Breakpoint> BreakpointManager::getBreakpointsForScript(int scriptId) const {
    std::vector<Breakpoint> result;
    for (const auto& [key, breakpoint] : breakpoints_) {
        (void)key;
        if (breakpoint.scriptId == scriptId) {
            result.push_back(breakpoint);
        }
    }
    return result;
}

std::set<int> BreakpointManager::getOffsetsForScript(int scriptId) const {
    std::set<int> offsets;
    for (const auto& [key, breakpoint] : breakpoints_) {
        (void)key;
        if (breakpoint.scriptId == scriptId) {
            offsets.insert(breakpoint.offset);
        }
    }
    return offsets;
}

void BreakpointManager::clearAll() {
    breakpoints_.clear();
}

std::map<int, std::set<int>> BreakpointManager::toOffsetMap() const {
    std::map<int, std::set<int>> result;
    for (const auto& [key, breakpoint] : breakpoints_) {
        (void)key;
        result[breakpoint.scriptId].insert(breakpoint.offset);
    }
    return result;
}

void BreakpointManager::setFromOffsetMap(const std::map<int, std::set<int>>& offsetMap) {
    clearAll();
    for (const auto& [scriptId, offsets] : offsetMap) {
        for (const int offset : offsets) {
            setBreakpoint(Breakpoint::simple(scriptId, "", offset));
        }
    }
}

std::string BreakpointManager::serialize() const {
    if (breakpoints_.empty()) {
        return "";
    }

    std::ostringstream out;
    out << "{\"version\":3,\"breakpoints\":[";
    bool first = true;
    for (const auto& [key, breakpoint] : breakpoints_) {
        (void)key;
        if (!first) {
            out << ",";
        }
        first = false;
        out << "{\"scriptId\":" << breakpoint.scriptId
            << ",\"handlerName\":\"" << escapeJson(breakpoint.handlerName)
            << "\",\"offset\":" << breakpoint.offset
            << ",\"enabled\":" << (breakpoint.enabled ? "true" : "false")
            << "}";
    }
    out << "]}";
    return out.str();
}

std::string BreakpointManager::escapeJson(const std::string& value) {
    std::string result;
    for (const char ch : value) {
        switch (ch) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            case '\r': result += "\\r"; break;
            case '\t': result += "\\t"; break;
            default: result += ch; break;
        }
    }
    return result;
}

void BreakpointManager::deserialize(std::string data) {
    clearAll();
    data = trim(std::move(data));
    if (data.empty()) {
        return;
    }
    if (data.front() == '{') {
        deserializeJson(data);
    } else {
        deserializeLegacy(data);
    }
}

void BreakpointManager::deserializeJson(const std::string& json) {
    const std::string marker = "\"breakpoints\":[";
    const auto breakpointsStart = json.find(marker);
    if (breakpointsStart == std::string::npos) {
        return;
    }
    const auto arrayStart = json.find('[', breakpointsStart);
    const auto arrayEnd = json.rfind(']');
    if (arrayStart == std::string::npos || arrayEnd == std::string::npos || arrayEnd <= arrayStart) {
        return;
    }

    const std::string arrayContent = json.substr(arrayStart + 1, arrayEnd - arrayStart - 1);
    if (trim(arrayContent).empty()) {
        return;
    }

    int depth = 0;
    std::size_t objectStart = std::string::npos;
    for (std::size_t index = 0; index < arrayContent.size(); ++index) {
        const char ch = arrayContent[index];
        if (ch == '{') {
            if (depth == 0) {
                objectStart = index;
            }
            ++depth;
        } else if (ch == '}') {
            --depth;
            if (depth == 0 && objectStart != std::string::npos) {
                auto breakpoint = parseBreakpointJson(arrayContent.substr(objectStart, index - objectStart + 1));
                if (breakpoint.has_value()) {
                    setBreakpoint(*breakpoint);
                }
                objectStart = std::string::npos;
            }
        }
    }
}

std::optional<Breakpoint> BreakpointManager::parseBreakpointJson(const std::string& json) {
    const int scriptId = parseJsonInt(json, "scriptId", 0);
    const std::string handlerName = parseJsonString(json, "handlerName", "");
    const int offset = parseJsonInt(json, "offset", 0);
    const bool enabled = parseJsonBoolean(json, "enabled", true);
    return Breakpoint{scriptId, handlerName, offset, enabled};
}

int BreakpointManager::parseJsonInt(const std::string& json, const std::string& key, int defaultValue) {
    const std::string pattern = "\"" + key + "\":";
    const auto index = json.find(pattern);
    if (index == std::string::npos) {
        return defaultValue;
    }

    std::size_t start = index + pattern.size();
    std::size_t end = start;
    while (end < json.size() && (std::isdigit(static_cast<unsigned char>(json[end])) || json[end] == '-')) {
        ++end;
    }
    if (end == start) {
        return defaultValue;
    }

    try {
        return std::stoi(json.substr(start, end - start));
    } catch (...) {
        return defaultValue;
    }
}

std::string BreakpointManager::parseJsonString(const std::string& json,
                                               const std::string& key,
                                               const std::string& defaultValue) {
    const std::string pattern = "\"" + key + "\":\"";
    const auto index = json.find(pattern);
    if (index == std::string::npos) {
        return defaultValue;
    }

    std::string result;
    bool escaped = false;
    for (std::size_t pos = index + pattern.size(); pos < json.size(); ++pos) {
        const char ch = json[pos];
        if (escaped) {
            switch (ch) {
                case '"': result += '"'; break;
                case '\\': result += '\\'; break;
                case 'n': result += '\n'; break;
                case 'r': result += '\r'; break;
                case 't': result += '\t'; break;
                default: result += ch; break;
            }
            escaped = false;
        } else if (ch == '\\') {
            escaped = true;
        } else if (ch == '"') {
            return result;
        } else {
            result += ch;
        }
    }
    return defaultValue;
}

bool BreakpointManager::parseJsonBoolean(const std::string& json,
                                         const std::string& key,
                                         bool defaultValue) {
    const std::string pattern = "\"" + key + "\":";
    const auto index = json.find(pattern);
    if (index == std::string::npos) {
        return defaultValue;
    }
    const std::size_t start = index + pattern.size();
    if (json.compare(start, 4, "true") == 0) {
        return true;
    }
    if (json.compare(start, 5, "false") == 0) {
        return false;
    }
    return defaultValue;
}

void BreakpointManager::deserializeLegacy(const std::string& data) {
    try {
        for (const auto& script : split(data, ';')) {
            if (script.empty()) {
                continue;
            }
            const auto parts = split(script, ':');
            if (parts.size() != 2) {
                continue;
            }
            const int scriptId = std::stoi(parts[0]);
            for (const auto& offsetText : split(parts[1], ',')) {
                if (!offsetText.empty()) {
                    setBreakpoint(Breakpoint::simple(scriptId, "", std::stoi(offsetText)));
                }
            }
        }
    } catch (...) {
    }
}

std::string BreakpointManager::serializeLegacy() const {
    const auto offsetMap = toOffsetMap();
    std::ostringstream out;
    bool firstScript = true;
    for (const auto& [scriptId, offsets] : offsetMap) {
        if (offsets.empty()) {
            continue;
        }
        if (!firstScript) {
            out << ";";
        }
        firstScript = false;
        out << scriptId << ":";
        bool firstOffset = true;
        for (const int offset : offsets) {
            if (!firstOffset) {
                out << ",";
            }
            firstOffset = false;
            out << offset;
        }
    }
    return out.str();
}

} // namespace libreshockwave::player::debug
